File: backends/open-liberty-java/src/main/java/dev/stackverse/openliberty/EventLogger.java
package dev.stackverse.openliberty;

import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.SpanContext;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.sql.SQLException;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;

/** Stackverse structured-event logger with CDI-managed configuration. */
@ApplicationScoped
public class EventLogger {
    @Inject AppConfig config;

    void event(String level, String event, String outcome, String message, Map<String, ?> fields) {
        if (!enabled(level)) {
            return;
        }
        Map<String, Object> line = new LinkedHashMap<>();
        line.put("timestamp", Instant.now().toString());
        line.put("level", level.toUpperCase());
        line.put("logger", "stackverse.open-liberty-java");
        line.put("message", message);
        line.put("event", event);
        line.put("outcome", outcome);
        SpanContext spanContext = Span.current().getSpanContext();
        if (spanContext.isValid()) {
            line.put("trace_id", spanContext.getTraceId());
            line.put("span_id", spanContext.getSpanId());
        }
        fields.forEach(line::put);
        if ("text".equals(config.logFormat())) {
            System.out.println(line);
        } else {
            System.out.println(JsonSupport.jsonString(line));
        }
    }

    void error(
            String event,
            String outcome,
            String message,
            Throwable throwable,
            Map<String, ?> fields) {
        failure("error", event, outcome, message, throwable, fields);
    }

    void fatal(
            String event,
            String outcome,
            String message,
            Throwable throwable,
            Map<String, ?> fields) {
        failure("fatal", event, outcome, message, throwable, fields);
    }

    private void failure(
            String level,
            String event,
            String outcome,
            String message,
            Throwable throwable,
            Map<String, ?> fields) {
        Map<String, Object> withError = new LinkedHashMap<>(fields);
        withError.put("stack_trace", stackTrace(throwable));
        event(level, event, outcome, message, withError);
    }

    void dependencyFailure(String dependency, Throwable throwable, long durationMs) {
        error(
                "dependency_call_failed",
                "failure",
                "Dependency call failed",
                throwable,
                Map.of(
                        "dependency", dependency,
                        "duration_ms", durationMs,
                        "error_code", errorCode(throwable)));
    }

    static boolean causedBySqlFailure(Throwable throwable) {
        for (Throwable current = throwable; current != null; current = current.getCause()) {
            if (current instanceof SQLException) {
                return true;
            }
        }
        return false;
    }

    static String errorCode(Throwable throwable) {
        Throwable root = throwable;
        while (root.getCause() != null) {
            root = root.getCause();
        }
        if (root instanceof SQLException sql && sql.getSQLState() != null) {
            return "sqlstate_" + sql.getSQLState().toLowerCase(Locale.ROOT);
        }
        return root.getClass().getSimpleName().toLowerCase(Locale.ROOT);
    }

    static long elapsedMillis(long startedAtNanos) {
        return Math.max(0L, (System.nanoTime() - startedAtNanos) / 1_000_000L);
    }

    private boolean enabled(String level) {
        return severity(level) <= severity(config.logLevel());
    }

    private static int severity(String level) {
        return switch (level.toLowerCase()) {
            case "error", "fatal" -> 0;
            case "warn" -> 1;
            case "debug" -> 3;
            default -> 2;
        };
    }

    private static String stackTrace(Throwable throwable) {
        StringWriter writer = new StringWriter();
        throwable.printStackTrace(new PrintWriter(writer));
        return writer.toString();
    }
}

File: backends/scala-http4s/src/main/scala/dev/stackverse/http4s/EventLogger.scala
package dev.stackverse.http4s

import io.circe.{Json, JsonObject}
import io.opentelemetry.api.common.AttributeKey
import io.opentelemetry.api.logs.{Logger as OtelLogger, Severity}
import io.opentelemetry.sdk.OpenTelemetrySdk
import io.opentelemetry.sdk.autoconfigure.AutoConfiguredOpenTelemetrySdk

import java.io.{PrintWriter, StringWriter}
import java.time.Instant
import java.time.format.DateTimeFormatter
import scala.util.Try

final class EventLogger(config: BackendConfig) extends ProblemEvents {
  private val priorities = Map("debug" -> 10, "info" -> 20, "warn" -> 30, "error" -> 40, "fatal" -> 50)
  private val threshold = priorities.getOrElse(config.logLevel, 20)
  private val otelSdk: Option[OpenTelemetrySdk] =
    if (config.otelEnabled) Some(AutoConfiguredOpenTelemetrySdk.initialize().getOpenTelemetrySdk) else None
  private val otelLogger: Option[OtelLogger] =
    otelSdk.map(_.getLogsBridge.loggerBuilder("stackverse-backend-scala-http4s").build())

  override def event(level: String, event: String, outcome: String, message: String, fields: (String, Json)*): Unit =
    write(level, message, Some(event), Some(outcome), fields*)

  def line(level: String, message: String, fields: (String, Json)*): Unit =
    write(level, message, None, None, fields*)

  def fatal(event: String, outcome: String, message: String, error: Throwable, fields: (String, Json)*): Unit = {
    val stack = StringWriter()
    error.printStackTrace(PrintWriter(stack))
    write(
      "fatal",
      message,
      Some(event),
      Some(outcome),
      (fields ++ Seq(
        "error_code" -> Json.fromString("application_start_failed"),
        "error_type" -> Json.fromString(error.getClass.getName),
        "stack_trace" -> Json.fromString(stack.toString)
      ))*
    )
  }

  private def write(
      level: String,
      message: String,
      eventName: Option[String],
      outcome: Option[String],
      fields: (String, Json)*
  ): Unit = {
    if (priorities.getOrElse(level, 20) < threshold) return
    val base = Seq(
      "timestamp" -> Json.fromString(DateTimeFormatter.ISO_INSTANT.format(Instant.now())),
      "level" -> Json.fromString(level),
      "message" -> Json.fromString(message)
    ) ++ eventName.map("event" -> Json.fromString(_)) ++ outcome.map("outcome" -> Json.fromString(_))
    val obj = JsonObject.fromIterable((base ++ fields).filterNot(_._2.isNull))
    if (config.logFormat == "text") {
      val suffix = fields.collect { case (key, value) if !value.isNull => s"$key=${value.noSpaces}" }
      println((Seq(level.toUpperCase, message) ++ suffix).mkString(" "))
    } else {
      println(Json.fromJsonObject(obj).noSpaces)
    }
    exportOtel(level, message, eventName, outcome, fields*)
  }

  def shutdown(): Unit =
    otelSdk.foreach(sdk => Try(sdk.close()))

  private def exportOtel(
      level: String,
      message: String,
      eventName: Option[String],
      outcome: Option[String],
      fields: (String, Json)*
  ): Unit =
    otelLogger.foreach { logger =>
      try {
        val record = logger
          .logRecordBuilder()
          .setTimestamp(Instant.now())
          .setSeverity(severity(level))
          .setBody(message)
        eventName.foreach(value => record.setAttribute(AttributeKey.stringKey("event"), value))
        outcome.foreach(value => record.setAttribute(AttributeKey.stringKey("outcome"), value))
        fields.foreach {
          case (key, value) if value.isString => record.setAttribute(AttributeKey.stringKey(key), value.asString.get)
          case (_, value) if value.isNull     => ()
          case (key, value)                   => record.setAttribute(AttributeKey.stringKey(key), value.noSpaces)
        }
        record.emit()
      } catch {
        case _: Throwable => ()
      }
    }

  private def severity(level: String): Severity = level match {
    case "debug" => Severity.DEBUG
    case "warn"  => Severity.WARN
    case "error" => Severity.ERROR
    case "fatal" => Severity.FATAL
    case _       => Severity.INFO
  }
}
